lib/filter-gmail-compiler.js

```diff
@@ -0,0 +1,131 @@
+// Compiler that transforms filter expression AST into a gmail search expression.
+//
+// The expression is returned from the compile method.
+
+var ASTVisitor = require('juttle/lib/compiler/ast-visitor');
+var JuttleErrors = require('juttle/lib/errors');
+var JuttleMoment = require('juttle/lib/moment').JuttleMoment;
+var _ = require('underscore');
+
+// FilterGmailCompiler derives from ASTVisitor which provides a way to
+// traverse the abstract syntax tree that the juttle compiler
+// generates for the read command's filter expression.
+//
+// While traversing the tree, callbacks are called for the various
+// parts of the filter expression. The FilterGmailCompiler object maps
+// individual items in the tree into appropriate gmail advanced search
+// (https://support.google.com/mail/answer/7190?hl=en) terms. As the
+// tree is traversed, these items are combined into a complete gmail
+// search expression. This is used when reading messages.
+
+var FilterGmailCompiler = ASTVisitor.extend({
+    initialize: function(options) {
+        this.location = options.location;
+        this.source = options.source;
+        this.supported_headers = ["from", "to", "subject", "cc", "bcc"];
+    },
+
+    compile: function(node) {
+        return this.visit(node);
+    },
+
+    visitStringLiteral: function(node) {
+        return node.value;
+    },
+
+    visitFilterLiteral: function(node) {
+        return this.visit(node.ast);
+    },
+
+    visitUnaryExpression: function(node) {
+        switch (node.operator) {
+            case 'NOT':
+                return '-' + this.visit(node.expression);
+
+            // '*' is the field dereferencing operator. For example,
+            // given a search string from ~ 'bob', the UnaryExpression
+            // * on from means "the field called from".
+            case '*':
+                return this.visit(node.expression);
+
+            default:
+                throw this.source.compile_error('RT-ADAPTER-UNSUPPORTED-FILTER',
+                                                {proc: 'read gmail', filter: "operator " + node.operator});
+        }
+    },
+
+    visitBinaryExpression: function(node) {
+        var left, right, filter;
+
+        switch (node.operator) {
+            case 'AND':
+                left = this.visit(node.left);
+                right = this.visit(node.right);
+
+                filter = "(" + left + " " + right + ")";
+                break;
+
+            case 'OR':
+                left = this.visit(node.left);
+                right = this.visit(node.right);
+
+                filter = "(" + left + " OR " + right + ")";
+
+                break;
+
+            // The gmail search syntax only supports substring matches
```

Comment:
Yes, they used to behave the same if no globbing characters were specified. And I agree that it would be annoying to force a person to specify them (I actually think that `~` should always to a substring match implicitly). Not quite proposing a solution here, but I think its important that adapters document situations in which they stray from conventions (and especially what is documented in the juttle docs).